mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/MapboxTelemetry.java

```diff
@@ -0,0 +1,573 @@
+   */
+  private boolean isPluggedIn() {
+    int chargePlug = batteryStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
+    if (chargePlug == BatteryManager.BATTERY_PLUGGED_USB
+      || chargePlug == BatteryManager.BATTERY_PLUGGED_AC) {
+      return true;
+    }
+
+    return false;
+  }
+
+  private void loadUserPreferences() {
+    SharedPreferences prefs = TelemetryUtils.getSharedPreferences(context);
+
+    // Load vendor ID
+    if (prefs.contains(TelemetryConstants.MAPBOX_SHARED_PREFERENCE_KEY_VENDOR_ID)) {
+      mapboxVendorId = prefs.getString(TelemetryConstants.MAPBOX_SHARED_PREFERENCE_KEY_VENDOR_ID, "");
+    }
+
+    // Create vendor ID (if needed)
+    if (TextUtils.isEmpty(mapboxVendorId)) {
+      mapboxVendorId = UUID.randomUUID().toString();
+      SharedPreferences.Editor editor = prefs.edit();
+      editor.putString(TelemetryConstants.MAPBOX_SHARED_PREFERENCE_KEY_VENDOR_ID, mapboxVendorId);
+      editor.apply();
+      editor.commit();
+    }
+
+    // Set telemetry opt-in/opt-out status
+    setTelemetryEnabled(isTelemetryEnabled());
+  }
+
+  public boolean isTelemetryEnabled() {
+    if (telemetryEnabled == null) {
+      // Cache value
+      telemetryEnabled = TelemetryUtils.getSharedPreferences(context)
+        .getBoolean(TelemetryConstants.MAPBOX_SHARED_PREFERENCE_KEY_TELEMETRY_ENABLED, true);
+    }
+
+    return telemetryEnabled;
+  }
+
+  /**
+   * Enables / disables telemetry
+   *
+   * @param telemetryEnabled True to start telemetry, false to stop it
+   */
+  public void setTelemetryEnabled(boolean telemetryEnabled) {
+    if (initialized && isTelemetryEnabled() == telemetryEnabled) {
+      Timber.v("Telemetry was already initialized on that state (enabled: %b).", telemetryEnabled);
+      return;
+    }
+
+    if (telemetryEnabled) {
+      Timber.v("Enabling telemetry.");
+      context.startService(new Intent(context, TelemetryService.class));
+
+      // Check for location permissions, periodically if necessary
+      if (PermissionsManager.areLocationPermissionsGranted(context)) {
+        registerLocationUpdates();
+      } else {
+        startPermissionsTimer();
+      }
+
+      // Send events timer
+      timer = new Timer();
+      timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          flushEventsQueueImmediately(false);
+        }
+      }, TelemetryConstants.FLUSH_DELAY_MS, TelemetryConstants.FLUSH_PERIOD_MS);
+    } else {
+      Timber.v("Disabling telemetry.");
+      withShutDown = true;
+
+      // This event is always recorded
+      pushTurnstileEvent();
+    }
+
+    // Persist
+    this.telemetryEnabled = telemetryEnabled;
+    SharedPreferences prefs = TelemetryUtils.getSharedPreferences(context);
+    SharedPreferences.Editor editor = prefs.edit();
+    editor.putBoolean(TelemetryConstants.MAPBOX_SHARED_PREFERENCE_KEY_TELEMETRY_ENABLED, telemetryEnabled);
+    editor.apply();
+    editor.commit();
```

Comment:
having both apply() and commit() is not needed they both result in saving, since this is called at application startup I would go for apply as this is asynchronous.